onnxruntime/core/optimizer/transformer_memcpy.cc

```diff
@@ -294,7 +294,26 @@ void TransformerMemcpyImpl::ProcessInitializers() {
   }
 
   for (auto p_node : provider_nodes_) {
-    p_node->ReplaceDefs(replacements);
+    auto dup_replacements = replacements;
+
+    const KernelCreateInfo* kci = nullptr;
+    kernel_registries.SearchKernelRegistry(*p_node, &kci);
+    p_node->ForEachWithIndex(
+        p_node->InputDefs(),
+        [p_node, kci, &dup_replacements](const onnxruntime::NodeArg& arg, size_t index) {
+          if (kci && MemTypeOnCpuExplicitly(kci->kernel_def->InputMemoryType(index)))
+            dup_replacements.erase(&arg);
+          return Status::OK();
+        });
+    p_node->ForEachWithIndex(
+        p_node->OutputDefs(),
+        [p_node, kci, &dup_replacements](const onnxruntime::NodeArg& arg, size_t index) {
```

Comment:
Why do we need to check OutputDefs? Wouldn't an initializer only ever appear as an input?

Reply:
Just to be safe, as we may have assign op later to modify initializer. Changing to assert instead.

---
In reply to: [265485136](https://github.com/Microsoft/onnxruntime/pull/623#discussion_r265485136) [](ancestors = 265485136)